add RPM variant to libabigail

woodard opened the pull request:
Add a variant to libabigail package which builds the RPM tools.  There
were two python dependencies missing py-xdg and py-koji and so I also
created packages for them.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
add RPM variant to libabigail

Add a variant to libabigail package which builds the RPM tools.  There
were two python dependencies missing py-xdg and py-koji and so I also
created packages for them.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard commented:
This is more of an RFC at the moment. It hasn't passed my local testing. For some reason the python include path is not getting set properly or something and so the tools are not able to find their dependencies.

```
$ fedabipkgdiff --self-compare -a --from fc36 libabigail
Traceback (most recent call last):
  File "/home/ben/Shared/Work/spack/opt/spack/linux-fedora36-skylake/gcc-12.1.1/libabigail-master-zgh7obpagd73icmgnrq3p27fadjyhzgm/bin/fedabipkgdiff", line 27, in <module>
    import xdg.BaseDirectory
ModuleNotFoundError: No module named 'xdg.BaseDirectory'
```

woodard committed:
Designate myself as maintainer

Add myself as the maintainer for the libabigail spack package.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard commented:
Another thing that I noticed is doing `spack load libabigail` does seem to put py-xdg from the spack build into the PYTHONPATH. This makes the error message that much more cryptic.

woodard committed:
Add more dependencies

Add rpm and cpio as dependencies to this. They exist on rpm based
systems but if you try to build on ubuntu it they will be needed.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Enable rpm's python library

fedabipkgdiff requires the python rpm library. This is a variant
available in the rpm recipe but the option was not wired up. Wired up
the config option and forced the variant in libabigail.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge remote-tracking branch 'origin/develop' into libabigail-rpm

woodard committed:
Add a couple more missing dependencies

For the koji python module to load properly two more dependencies were
needed py-python-dateutil and py-requests.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Fix a typo

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Move package dependencies to the right place

I realized that the dependencies for py-koji needed to be in the
py-koji package not in the libabigail package.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Forgot to take off the when

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge remote-tracking branch 'origin/develop' into libabigail-rpm

woodard committed:
Use the right pypi package

The name of the package that fedabipkgdiff needs is not xdg it is
pyxdg. So I removed the new module that I made in a previous
commit. Then added the correct new one.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge remote-tracking branch 'origin' into libabigail-rpm

woodard committed:
Make RPM extend python when the option is configured.

The RPM package can provide python, this is needed for
fedabipkgdiff. However, when I wired up the option to add python to
the RPM package I forgot to say that it extends python.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Fix the way that I extend python

Didn't have the spack syntax for extending done correctly.

Signed-off-by: Ben Woodard <woodard@redhat.com>

adamjstewart commented on var/spack/repos/builtin/packages/py-koji/package.py:
```diff
@@ -0,0 +1,20 @@
+
+    extends('python', ignore=r'bin/pytest')
+    depends_on('py-python-dateutil', type=('build', 'run'))
+    depends_on('py-requests', type=('build', 'run'))
```
Missing deps on requests-gssapi and six

woodard replied:
I do not see where requests-gssapi is required.

adamjstewart replied:
In the `setup.py` file. If you're unfamiliar with Python's build system, https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html is a helpful read. This package uses setuptools so you'll want to pay close attention to the setuptools section.

adamjstewart commented on var/spack/repos/builtin/packages/py-koji/package.py:
```diff
@@ -0,0 +1,20 @@
+    extends('python', ignore=r'bin/pytest')
+    depends_on('py-python-dateutil', type=('build', 'run'))
+    depends_on('py-requests', type=('build', 'run'))
+    depends_on('python@3.6:', type=('build', 'run'))
```
```suggestion
    depends_on('python@2.7:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pyxdg/package.py:
```diff
@@ -0,0 +1,19 @@
+    version('0.28', sha256='3267bb3074e934df202af2ee0868575484108581e6f3cb006af1da35395e88b4')
+
+    extends('python', ignore=r'bin/pytest')
+    depends_on('python@3.6:', type=('build', 'run'))
```
Where do you see a 3.6+ requirement? I think you can remove this line

adamjstewart commented on var/spack/repos/builtin/packages/py-pyxdg/package.py:
```diff
@@ -0,0 +1,19 @@
+
+    extends('python', ignore=r'bin/pytest')
+    depends_on('python@3.6:', type=('build', 'run'))
+#    depends_on('py-poetry-core@1.0.0:', type='build')
```
This line can be removed

adamjstewart commented on var/spack/repos/builtin/packages/py-koji/package.py:
```diff
@@ -0,0 +1,20 @@
+
+    version('1.29.0', sha256='7db1ea4f68dd94d556c612e40bbfc2268164bfebc4a4cd08b63971bddaa686bc')
+
+    extends('python', ignore=r'bin/pytest')
```
Is this actually needed? I think this only matters when using `spack activate` which we are in the process of removing.

woodard replied:
I don't follow you on this one.

adamjstewart replied:
I don't think `ignore='bin/pytest'` should be needed. Why did you add it?

adamjstewart reviewed (changes requested):
Will try building locally to see why you're getting that error message.

adamjstewart commented:
To debug this I would:

1. Uninstall and reinstall these 4 packages, maybe you have an old copy lying around
2. Run `spack load libabigail` again
3. Check `echo $PYTHONPATH` and make sure it contains `py-pyxdg`
4. Try the same import yourself in Python
5. Make sure you don't have an `xdg` directory in your current directory that is masking the import

FWIW, this worked for me:
```console
$ spack load py-pyxdg
$ python
Python 3.9.13 (main, Jun 21 2022, 22:34:57) 
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import xdg.BaseDirectory
>>>
```

woodard committed:
Merge remote-tracking branch 'origin/develop' into libabigail-rpm

woodard committed:
Fix up dependencies on python packages

Found out how to track down the dependencies for python packages and
so I replaced boilerplate code with the correct values.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge remote-tracking branch 'origin/develop' into libabigail-rpm

woodard committed:
Exclude MacOS from RPM build targets.

RPM is a Linux tool and it has deep assunmptions about the Linux file
system hierarchy which are not satisfied by MacOS, which is only Unix
like. Even if could be made to compile, it would not run
correctly. Plus MacOS has its own package management system which is
different than RPM and so no one really wants RPM on MacOS.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge branch 'develop' into libabigail-rpm

alalazo commented on var/spack/repos/builtin/packages/libabigail/package.py:
```diff
@@ -12,12 +12,14 @@ class Libabigail(AutotoolsPackage):
     version('1.8', sha256='1cbf260b894ccafc61b2673ba30c020c3f67dbba9dfa88dca3935dff661d665c')
 
     variant('docs', default=False, description='build documentation')
+    variant('rpm', default=False, description='build RPM tools')
```
Would it make sense to have this variant only on linux::
```suggestion
    variant('rpm', default=False, when='platform=linux', description='build RPM tools')
```
?

alalazo commented:
@woodard

alecbcs commented:
Going to go head and close this PR due to a lack of response. @woodard happy to reopen if you still need this variant added.
